// cosmjs-types/cosmos/benchmark/v1/tx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgClientImpl = exports.MsgLoadTestResponse = exports.MsgLoadTest = exports.protobufPackage = void 0;
/* eslint-disable */
const benchmark_1 = require("./benchmark");
const binary_1 = require("../../../binary");
const helpers_1 = require("../../../helpers");
exports.protobufPackage = "cosmos.benchmark.v1";
function createBaseMsgLoadTest() {
    return {
        caller: new Uint8Array(),
        ops: [],
    };
}
exports.MsgLoadTest = {
    typeUrl: "/cosmos.benchmark.v1.MsgLoadTest",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.caller.length !== 0) {
            writer.uint32(10).bytes(message.caller);
        }
        for (const v of message.ops) {
            benchmark_1.Op.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgLoadTest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.caller = reader.bytes();
                    break;
                case 2:
                    message.ops.push(benchmark_1.Op.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseMsgLoadTest();
        if ((0, helpers_1.isSet)(object.caller))
            obj.caller = (0, helpers_1.bytesFromBase64)(object.caller);
        if (Array.isArray(object?.ops))
            obj.ops = object.ops.map((e) => benchmark_1.Op.fromJSON(e));
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.caller !== undefined &&
            (obj.caller = (0, helpers_1.base64FromBytes)(message.caller !== undefined ? message.caller : new Uint8Array()));
        if (message.ops) {
            obj.ops = message.ops.map((e) => (e ? benchmark_1.Op.toJSON(e) : undefined));
        }
        else {
            obj.ops = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseMsgLoadTest();
        message.caller = object.caller ?? new Uint8Array();
        message.ops = object.ops?.map((e) => benchmark_1.Op.fromPartial(e)) || [];
        return message;
    },
};
function createBaseMsgLoadTestResponse() {
    return {
        totalTime: BigInt(0),
        totalErrors: BigInt(0),
    };
}
exports.MsgLoadTestResponse = {
    typeUrl: "/cosmos.benchmark.v1.MsgLoadTestResponse",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.totalTime !== BigInt(0)) {
            writer.uint32(8).uint64(message.totalTime);
        }
        if (message.totalErrors !== BigInt(0)) {
            writer.uint32(16).uint64(message.totalErrors);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgLoadTestResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.totalTime = reader.uint64();
                    break;
                case 2:
                    message.totalErrors = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseMsgLoadTestResponse();
        if ((0, helpers_1.isSet)(object.totalTime))
            obj.totalTime = BigInt(object.totalTime.toString());
        if ((0, helpers_1.isSet)(object.totalErrors))
            obj.totalErrors = BigInt(object.totalErrors.toString());
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.totalTime !== undefined && (obj.totalTime = (message.totalTime || BigInt(0)).toString());
        message.totalErrors !== undefined && (obj.totalErrors = (message.totalErrors || BigInt(0)).toString());
        return obj;
    },
    fromPartial(object) {
        const message = createBaseMsgLoadTestResponse();
        if (object.totalTime !== undefined && object.totalTime !== null) {
            message.totalTime = BigInt(object.totalTime.toString());
        }
        if (object.totalErrors !== undefined && object.totalErrors !== null) {
            message.totalErrors = BigInt(object.totalErrors.toString());
        }
        return message;
    },
};
class MsgClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
        this.LoadTest = this.LoadTest.bind(this);
    }
    LoadTest(request) {
        const data = exports.MsgLoadTest.encode(request).finish();
        const promise = this.rpc.request("cosmos.benchmark.v1.Msg", "LoadTest", data);
        return promise.then((data) => exports.MsgLoadTestResponse.decode(new binary_1.BinaryReader(data)));
    }
}
exports.MsgClientImpl = MsgClientImpl;
