// cosmjs-types/cosmos/protocolpool/v1/genesis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
const types_1 = require("./types");
const binary_1 = require("../../../binary");
const helpers_1 = require("../../../helpers");
exports.protobufPackage = "cosmos.protocolpool.v1";
function createBaseGenesisState() {
    return {
        continuousFunds: [],
        params: types_1.Params.fromPartial({}),
    };
}
exports.GenesisState = {
    typeUrl: "/cosmos.protocolpool.v1.GenesisState",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        for (const v of message.continuousFunds) {
            types_1.ContinuousFund.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.params !== undefined) {
            types_1.Params.encode(message.params, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.continuousFunds.push(types_1.ContinuousFund.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.params = types_1.Params.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseGenesisState();
        if (Array.isArray(object?.continuousFunds))
            obj.continuousFunds = object.continuousFunds.map((e) => types_1.ContinuousFund.fromJSON(e));
        if ((0, helpers_1.isSet)(object.params))
            obj.params = types_1.Params.fromJSON(object.params);
        return obj;
    },
    toJSON(message) {
        const obj = {};
        if (message.continuousFunds) {
            obj.continuousFunds = message.continuousFunds.map((e) => (e ? types_1.ContinuousFund.toJSON(e) : undefined));
        }
        else {
            obj.continuousFunds = [];
        }
        message.params !== undefined && (obj.params = message.params ? types_1.Params.toJSON(message.params) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGenesisState();
        message.continuousFunds = object.continuousFunds?.map((e) => types_1.ContinuousFund.fromPartial(e)) || [];
        if (object.params !== undefined && object.params !== null) {
            message.params = types_1.Params.fromPartial(object.params);
        }
        return message;
    },
};
