// cosmjs-types/ibc/core/channel/v2/genesis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketSequence = exports.PacketState = exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
const binary_1 = require("../../../../binary");
const helpers_1 = require("../../../../helpers");
exports.protobufPackage = "ibc.core.channel.v2";
function createBaseGenesisState() {
    return {
        acknowledgements: [],
        commitments: [],
        receipts: [],
        asyncPackets: [],
        sendSequences: [],
    };
}
exports.GenesisState = {
    typeUrl: "/ibc.core.channel.v2.GenesisState",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        for (const v of message.acknowledgements) {
            exports.PacketState.encode(v, writer.uint32(18).fork()).ldelim();
        }
        for (const v of message.commitments) {
            exports.PacketState.encode(v, writer.uint32(26).fork()).ldelim();
        }
        for (const v of message.receipts) {
            exports.PacketState.encode(v, writer.uint32(34).fork()).ldelim();
        }
        for (const v of message.asyncPackets) {
            exports.PacketState.encode(v, writer.uint32(42).fork()).ldelim();
        }
        for (const v of message.sendSequences) {
            exports.PacketSequence.encode(v, writer.uint32(50).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 2:
                    message.acknowledgements.push(exports.PacketState.decode(reader, reader.uint32()));
                    break;
                case 3:
                    message.commitments.push(exports.PacketState.decode(reader, reader.uint32()));
                    break;
                case 4:
                    message.receipts.push(exports.PacketState.decode(reader, reader.uint32()));
                    break;
                case 5:
                    message.asyncPackets.push(exports.PacketState.decode(reader, reader.uint32()));
                    break;
                case 6:
                    message.sendSequences.push(exports.PacketSequence.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseGenesisState();
        if (Array.isArray(object?.acknowledgements))
            obj.acknowledgements = object.acknowledgements.map((e) => exports.PacketState.fromJSON(e));
        if (Array.isArray(object?.commitments))
            obj.commitments = object.commitments.map((e) => exports.PacketState.fromJSON(e));
        if (Array.isArray(object?.receipts))
            obj.receipts = object.receipts.map((e) => exports.PacketState.fromJSON(e));
        if (Array.isArray(object?.asyncPackets))
            obj.asyncPackets = object.asyncPackets.map((e) => exports.PacketState.fromJSON(e));
        if (Array.isArray(object?.sendSequences))
            obj.sendSequences = object.sendSequences.map((e) => exports.PacketSequence.fromJSON(e));
        return obj;
    },
    toJSON(message) {
        const obj = {};
        if (message.acknowledgements) {
            obj.acknowledgements = message.acknowledgements.map((e) => (e ? exports.PacketState.toJSON(e) : undefined));
        }
        else {
            obj.acknowledgements = [];
        }
        if (message.commitments) {
            obj.commitments = message.commitments.map((e) => (e ? exports.PacketState.toJSON(e) : undefined));
        }
        else {
            obj.commitments = [];
        }
        if (message.receipts) {
            obj.receipts = message.receipts.map((e) => (e ? exports.PacketState.toJSON(e) : undefined));
        }
        else {
            obj.receipts = [];
        }
        if (message.asyncPackets) {
            obj.asyncPackets = message.asyncPackets.map((e) => (e ? exports.PacketState.toJSON(e) : undefined));
        }
        else {
            obj.asyncPackets = [];
        }
        if (message.sendSequences) {
            obj.sendSequences = message.sendSequences.map((e) => (e ? exports.PacketSequence.toJSON(e) : undefined));
        }
        else {
            obj.sendSequences = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGenesisState();
        message.acknowledgements = object.acknowledgements?.map((e) => exports.PacketState.fromPartial(e)) || [];
        message.commitments = object.commitments?.map((e) => exports.PacketState.fromPartial(e)) || [];
        message.receipts = object.receipts?.map((e) => exports.PacketState.fromPartial(e)) || [];
        message.asyncPackets = object.asyncPackets?.map((e) => exports.PacketState.fromPartial(e)) || [];
        message.sendSequences = object.sendSequences?.map((e) => exports.PacketSequence.fromPartial(e)) || [];
        return message;
    },
};
function createBasePacketState() {
    return {
        clientId: "",
        sequence: BigInt(0),
        data: new Uint8Array(),
    };
}
exports.PacketState = {
    typeUrl: "/ibc.core.channel.v2.PacketState",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.clientId !== "") {
            writer.uint32(10).string(message.clientId);
        }
        if (message.sequence !== BigInt(0)) {
            writer.uint32(16).uint64(message.sequence);
        }
        if (message.data.length !== 0) {
            writer.uint32(26).bytes(message.data);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePacketState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.clientId = reader.string();
                    break;
                case 2:
                    message.sequence = reader.uint64();
                    break;
                case 3:
                    message.data = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBasePacketState();
        if ((0, helpers_1.isSet)(object.clientId))
            obj.clientId = String(object.clientId);
        if ((0, helpers_1.isSet)(object.sequence))
            obj.sequence = BigInt(object.sequence.toString());
        if ((0, helpers_1.isSet)(object.data))
            obj.data = (0, helpers_1.bytesFromBase64)(object.data);
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.clientId !== undefined && (obj.clientId = message.clientId);
        message.sequence !== undefined && (obj.sequence = (message.sequence || BigInt(0)).toString());
        message.data !== undefined &&
            (obj.data = (0, helpers_1.base64FromBytes)(message.data !== undefined ? message.data : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        const message = createBasePacketState();
        message.clientId = object.clientId ?? "";
        if (object.sequence !== undefined && object.sequence !== null) {
            message.sequence = BigInt(object.sequence.toString());
        }
        message.data = object.data ?? new Uint8Array();
        return message;
    },
};
function createBasePacketSequence() {
    return {
        clientId: "",
        sequence: BigInt(0),
    };
}
exports.PacketSequence = {
    typeUrl: "/ibc.core.channel.v2.PacketSequence",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.clientId !== "") {
            writer.uint32(10).string(message.clientId);
        }
        if (message.sequence !== BigInt(0)) {
            writer.uint32(16).uint64(message.sequence);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePacketSequence();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.clientId = reader.string();
                    break;
                case 2:
                    message.sequence = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBasePacketSequence();
        if ((0, helpers_1.isSet)(object.clientId))
            obj.clientId = String(object.clientId);
        if ((0, helpers_1.isSet)(object.sequence))
            obj.sequence = BigInt(object.sequence.toString());
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.clientId !== undefined && (obj.clientId = message.clientId);
        message.sequence !== undefined && (obj.sequence = (message.sequence || BigInt(0)).toString());
        return obj;
    },
    fromPartial(object) {
        const message = createBasePacketSequence();
        message.clientId = object.clientId ?? "";
        if (object.sequence !== undefined && object.sequence !== null) {
            message.sequence = BigInt(object.sequence.toString());
        }
        return message;
    },
};
