// cosmjs-types/ibc/core/channel/v2/packet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecvPacketResult = exports.Acknowledgement = exports.Payload = exports.Packet = exports.PacketStatus = exports.protobufPackage = void 0;
exports.packetStatusFromJSON = packetStatusFromJSON;
exports.packetStatusToJSON = packetStatusToJSON;
/* eslint-disable */
const binary_1 = require("../../../../binary");
const helpers_1 = require("../../../../helpers");
exports.protobufPackage = "ibc.core.channel.v2";
/** PacketStatus specifies the status of a RecvPacketResult. */
var PacketStatus;
(function (PacketStatus) {
    /** PACKET_STATUS_UNSPECIFIED - PACKET_STATUS_UNSPECIFIED indicates an unknown packet status. */
    PacketStatus[PacketStatus["PACKET_STATUS_UNSPECIFIED"] = 0] = "PACKET_STATUS_UNSPECIFIED";
    /** PACKET_STATUS_SUCCESS - PACKET_STATUS_SUCCESS indicates a successful packet receipt. */
    PacketStatus[PacketStatus["PACKET_STATUS_SUCCESS"] = 1] = "PACKET_STATUS_SUCCESS";
    /** PACKET_STATUS_FAILURE - PACKET_STATUS_FAILURE indicates a failed packet receipt. */
    PacketStatus[PacketStatus["PACKET_STATUS_FAILURE"] = 2] = "PACKET_STATUS_FAILURE";
    /** PACKET_STATUS_ASYNC - PACKET_STATUS_ASYNC indicates an async packet receipt. */
    PacketStatus[PacketStatus["PACKET_STATUS_ASYNC"] = 3] = "PACKET_STATUS_ASYNC";
    PacketStatus[PacketStatus["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(PacketStatus || (exports.PacketStatus = PacketStatus = {}));
function packetStatusFromJSON(object) {
    switch (object) {
        case 0:
        case "PACKET_STATUS_UNSPECIFIED":
            return PacketStatus.PACKET_STATUS_UNSPECIFIED;
        case 1:
        case "PACKET_STATUS_SUCCESS":
            return PacketStatus.PACKET_STATUS_SUCCESS;
        case 2:
        case "PACKET_STATUS_FAILURE":
            return PacketStatus.PACKET_STATUS_FAILURE;
        case 3:
        case "PACKET_STATUS_ASYNC":
            return PacketStatus.PACKET_STATUS_ASYNC;
        case -1:
        case "UNRECOGNIZED":
        default:
            return PacketStatus.UNRECOGNIZED;
    }
}
function packetStatusToJSON(object) {
    switch (object) {
        case PacketStatus.PACKET_STATUS_UNSPECIFIED:
            return "PACKET_STATUS_UNSPECIFIED";
        case PacketStatus.PACKET_STATUS_SUCCESS:
            return "PACKET_STATUS_SUCCESS";
        case PacketStatus.PACKET_STATUS_FAILURE:
            return "PACKET_STATUS_FAILURE";
        case PacketStatus.PACKET_STATUS_ASYNC:
            return "PACKET_STATUS_ASYNC";
        case PacketStatus.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBasePacket() {
    return {
        sequence: BigInt(0),
        sourceClient: "",
        destinationClient: "",
        timeoutTimestamp: BigInt(0),
        payloads: [],
    };
}
exports.Packet = {
    typeUrl: "/ibc.core.channel.v2.Packet",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.sequence !== BigInt(0)) {
            writer.uint32(8).uint64(message.sequence);
        }
        if (message.sourceClient !== "") {
            writer.uint32(18).string(message.sourceClient);
        }
        if (message.destinationClient !== "") {
            writer.uint32(26).string(message.destinationClient);
        }
        if (message.timeoutTimestamp !== BigInt(0)) {
            writer.uint32(32).uint64(message.timeoutTimestamp);
        }
        for (const v of message.payloads) {
            exports.Payload.encode(v, writer.uint32(42).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePacket();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sequence = reader.uint64();
                    break;
                case 2:
                    message.sourceClient = reader.string();
                    break;
                case 3:
                    message.destinationClient = reader.string();
                    break;
                case 4:
                    message.timeoutTimestamp = reader.uint64();
                    break;
                case 5:
                    message.payloads.push(exports.Payload.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBasePacket();
        if ((0, helpers_1.isSet)(object.sequence))
            obj.sequence = BigInt(object.sequence.toString());
        if ((0, helpers_1.isSet)(object.sourceClient))
            obj.sourceClient = String(object.sourceClient);
        if ((0, helpers_1.isSet)(object.destinationClient))
            obj.destinationClient = String(object.destinationClient);
        if ((0, helpers_1.isSet)(object.timeoutTimestamp))
            obj.timeoutTimestamp = BigInt(object.timeoutTimestamp.toString());
        if (Array.isArray(object?.payloads))
            obj.payloads = object.payloads.map((e) => exports.Payload.fromJSON(e));
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.sequence !== undefined && (obj.sequence = (message.sequence || BigInt(0)).toString());
        message.sourceClient !== undefined && (obj.sourceClient = message.sourceClient);
        message.destinationClient !== undefined && (obj.destinationClient = message.destinationClient);
        message.timeoutTimestamp !== undefined &&
            (obj.timeoutTimestamp = (message.timeoutTimestamp || BigInt(0)).toString());
        if (message.payloads) {
            obj.payloads = message.payloads.map((e) => (e ? exports.Payload.toJSON(e) : undefined));
        }
        else {
            obj.payloads = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBasePacket();
        if (object.sequence !== undefined && object.sequence !== null) {
            message.sequence = BigInt(object.sequence.toString());
        }
        message.sourceClient = object.sourceClient ?? "";
        message.destinationClient = object.destinationClient ?? "";
        if (object.timeoutTimestamp !== undefined && object.timeoutTimestamp !== null) {
            message.timeoutTimestamp = BigInt(object.timeoutTimestamp.toString());
        }
        message.payloads = object.payloads?.map((e) => exports.Payload.fromPartial(e)) || [];
        return message;
    },
};
function createBasePayload() {
    return {
        sourcePort: "",
        destinationPort: "",
        version: "",
        encoding: "",
        value: new Uint8Array(),
    };
}
exports.Payload = {
    typeUrl: "/ibc.core.channel.v2.Payload",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.sourcePort !== "") {
            writer.uint32(10).string(message.sourcePort);
        }
        if (message.destinationPort !== "") {
            writer.uint32(18).string(message.destinationPort);
        }
        if (message.version !== "") {
            writer.uint32(26).string(message.version);
        }
        if (message.encoding !== "") {
            writer.uint32(34).string(message.encoding);
        }
        if (message.value.length !== 0) {
            writer.uint32(42).bytes(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePayload();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sourcePort = reader.string();
                    break;
                case 2:
                    message.destinationPort = reader.string();
                    break;
                case 3:
                    message.version = reader.string();
                    break;
                case 4:
                    message.encoding = reader.string();
                    break;
                case 5:
                    message.value = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBasePayload();
        if ((0, helpers_1.isSet)(object.sourcePort))
            obj.sourcePort = String(object.sourcePort);
        if ((0, helpers_1.isSet)(object.destinationPort))
            obj.destinationPort = String(object.destinationPort);
        if ((0, helpers_1.isSet)(object.version))
            obj.version = String(object.version);
        if ((0, helpers_1.isSet)(object.encoding))
            obj.encoding = String(object.encoding);
        if ((0, helpers_1.isSet)(object.value))
            obj.value = (0, helpers_1.bytesFromBase64)(object.value);
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.sourcePort !== undefined && (obj.sourcePort = message.sourcePort);
        message.destinationPort !== undefined && (obj.destinationPort = message.destinationPort);
        message.version !== undefined && (obj.version = message.version);
        message.encoding !== undefined && (obj.encoding = message.encoding);
        message.value !== undefined &&
            (obj.value = (0, helpers_1.base64FromBytes)(message.value !== undefined ? message.value : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        const message = createBasePayload();
        message.sourcePort = object.sourcePort ?? "";
        message.destinationPort = object.destinationPort ?? "";
        message.version = object.version ?? "";
        message.encoding = object.encoding ?? "";
        message.value = object.value ?? new Uint8Array();
        return message;
    },
};
function createBaseAcknowledgement() {
    return {
        appAcknowledgements: [],
    };
}
exports.Acknowledgement = {
    typeUrl: "/ibc.core.channel.v2.Acknowledgement",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        for (const v of message.appAcknowledgements) {
            writer.uint32(10).bytes(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAcknowledgement();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.appAcknowledgements.push(reader.bytes());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseAcknowledgement();
        if (Array.isArray(object?.appAcknowledgements))
            obj.appAcknowledgements = object.appAcknowledgements.map((e) => (0, helpers_1.bytesFromBase64)(e));
        return obj;
    },
    toJSON(message) {
        const obj = {};
        if (message.appAcknowledgements) {
            obj.appAcknowledgements = message.appAcknowledgements.map((e) => (0, helpers_1.base64FromBytes)(e !== undefined ? e : new Uint8Array()));
        }
        else {
            obj.appAcknowledgements = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAcknowledgement();
        message.appAcknowledgements = object.appAcknowledgements?.map((e) => e) || [];
        return message;
    },
};
function createBaseRecvPacketResult() {
    return {
        status: 0,
        acknowledgement: new Uint8Array(),
    };
}
exports.RecvPacketResult = {
    typeUrl: "/ibc.core.channel.v2.RecvPacketResult",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.status !== 0) {
            writer.uint32(8).int32(message.status);
        }
        if (message.acknowledgement.length !== 0) {
            writer.uint32(18).bytes(message.acknowledgement);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRecvPacketResult();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.status = reader.int32();
                    break;
                case 2:
                    message.acknowledgement = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseRecvPacketResult();
        if ((0, helpers_1.isSet)(object.status))
            obj.status = packetStatusFromJSON(object.status);
        if ((0, helpers_1.isSet)(object.acknowledgement))
            obj.acknowledgement = (0, helpers_1.bytesFromBase64)(object.acknowledgement);
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.status !== undefined && (obj.status = packetStatusToJSON(message.status));
        message.acknowledgement !== undefined &&
            (obj.acknowledgement = (0, helpers_1.base64FromBytes)(message.acknowledgement !== undefined ? message.acknowledgement : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        const message = createBaseRecvPacketResult();
        message.status = object.status ?? 0;
        message.acknowledgement = object.acknowledgement ?? new Uint8Array();
        return message;
    },
};
