// cosmjs-types/cosmos/benchmark/module/v1/module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeneratorParams = exports.Module = exports.protobufPackage = void 0;
/* eslint-disable */
const binary_1 = require("../../../../binary");
const helpers_1 = require("../../../../helpers");
exports.protobufPackage = "cosmos.benchmark.module.v1";
function createBaseModule() {
    return {
        genesisParams: undefined,
    };
}
exports.Module = {
    typeUrl: "/cosmos.benchmark.module.v1.Module",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.genesisParams !== undefined) {
            exports.GeneratorParams.encode(message.genesisParams, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseModule();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.genesisParams = exports.GeneratorParams.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseModule();
        if ((0, helpers_1.isSet)(object.genesisParams))
            obj.genesisParams = exports.GeneratorParams.fromJSON(object.genesisParams);
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.genesisParams !== undefined &&
            (obj.genesisParams = message.genesisParams ? exports.GeneratorParams.toJSON(message.genesisParams) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseModule();
        if (object.genesisParams !== undefined && object.genesisParams !== null) {
            message.genesisParams = exports.GeneratorParams.fromPartial(object.genesisParams);
        }
        return message;
    },
};
function createBaseGeneratorParams() {
    return {
        seed: BigInt(0),
        bucketCount: BigInt(0),
        keyMean: BigInt(0),
        keyStdDev: BigInt(0),
        valueMean: BigInt(0),
        valueStdDev: BigInt(0),
        genesisCount: BigInt(0),
        insertWeight: 0,
        updateWeight: 0,
        getWeight: 0,
        deleteWeight: 0,
    };
}
exports.GeneratorParams = {
    typeUrl: "/cosmos.benchmark.module.v1.GeneratorParams",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.seed !== BigInt(0)) {
            writer.uint32(8).uint64(message.seed);
        }
        if (message.bucketCount !== BigInt(0)) {
            writer.uint32(16).uint64(message.bucketCount);
        }
        if (message.keyMean !== BigInt(0)) {
            writer.uint32(24).uint64(message.keyMean);
        }
        if (message.keyStdDev !== BigInt(0)) {
            writer.uint32(32).uint64(message.keyStdDev);
        }
        if (message.valueMean !== BigInt(0)) {
            writer.uint32(48).uint64(message.valueMean);
        }
        if (message.valueStdDev !== BigInt(0)) {
            writer.uint32(56).uint64(message.valueStdDev);
        }
        if (message.genesisCount !== BigInt(0)) {
            writer.uint32(64).uint64(message.genesisCount);
        }
        if (message.insertWeight !== 0) {
            writer.uint32(77).float(message.insertWeight);
        }
        if (message.updateWeight !== 0) {
            writer.uint32(85).float(message.updateWeight);
        }
        if (message.getWeight !== 0) {
            writer.uint32(101).float(message.getWeight);
        }
        if (message.deleteWeight !== 0) {
            writer.uint32(93).float(message.deleteWeight);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGeneratorParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.seed = reader.uint64();
                    break;
                case 2:
                    message.bucketCount = reader.uint64();
                    break;
                case 3:
                    message.keyMean = reader.uint64();
                    break;
                case 4:
                    message.keyStdDev = reader.uint64();
                    break;
                case 6:
                    message.valueMean = reader.uint64();
                    break;
                case 7:
                    message.valueStdDev = reader.uint64();
                    break;
                case 8:
                    message.genesisCount = reader.uint64();
                    break;
                case 9:
                    message.insertWeight = reader.float();
                    break;
                case 10:
                    message.updateWeight = reader.float();
                    break;
                case 12:
                    message.getWeight = reader.float();
                    break;
                case 11:
                    message.deleteWeight = reader.float();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseGeneratorParams();
        if ((0, helpers_1.isSet)(object.seed))
            obj.seed = BigInt(object.seed.toString());
        if ((0, helpers_1.isSet)(object.bucketCount))
            obj.bucketCount = BigInt(object.bucketCount.toString());
        if ((0, helpers_1.isSet)(object.keyMean))
            obj.keyMean = BigInt(object.keyMean.toString());
        if ((0, helpers_1.isSet)(object.keyStdDev))
            obj.keyStdDev = BigInt(object.keyStdDev.toString());
        if ((0, helpers_1.isSet)(object.valueMean))
            obj.valueMean = BigInt(object.valueMean.toString());
        if ((0, helpers_1.isSet)(object.valueStdDev))
            obj.valueStdDev = BigInt(object.valueStdDev.toString());
        if ((0, helpers_1.isSet)(object.genesisCount))
            obj.genesisCount = BigInt(object.genesisCount.toString());
        if ((0, helpers_1.isSet)(object.insertWeight))
            obj.insertWeight = Number(object.insertWeight);
        if ((0, helpers_1.isSet)(object.updateWeight))
            obj.updateWeight = Number(object.updateWeight);
        if ((0, helpers_1.isSet)(object.getWeight))
            obj.getWeight = Number(object.getWeight);
        if ((0, helpers_1.isSet)(object.deleteWeight))
            obj.deleteWeight = Number(object.deleteWeight);
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.seed !== undefined && (obj.seed = (message.seed || BigInt(0)).toString());
        message.bucketCount !== undefined && (obj.bucketCount = (message.bucketCount || BigInt(0)).toString());
        message.keyMean !== undefined && (obj.keyMean = (message.keyMean || BigInt(0)).toString());
        message.keyStdDev !== undefined && (obj.keyStdDev = (message.keyStdDev || BigInt(0)).toString());
        message.valueMean !== undefined && (obj.valueMean = (message.valueMean || BigInt(0)).toString());
        message.valueStdDev !== undefined && (obj.valueStdDev = (message.valueStdDev || BigInt(0)).toString());
        message.genesisCount !== undefined && (obj.genesisCount = (message.genesisCount || BigInt(0)).toString());
        message.insertWeight !== undefined && (obj.insertWeight = message.insertWeight);
        message.updateWeight !== undefined && (obj.updateWeight = message.updateWeight);
        message.getWeight !== undefined && (obj.getWeight = message.getWeight);
        message.deleteWeight !== undefined && (obj.deleteWeight = message.deleteWeight);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGeneratorParams();
        if (object.seed !== undefined && object.seed !== null) {
            message.seed = BigInt(object.seed.toString());
        }
        if (object.bucketCount !== undefined && object.bucketCount !== null) {
            message.bucketCount = BigInt(object.bucketCount.toString());
        }
        if (object.keyMean !== undefined && object.keyMean !== null) {
            message.keyMean = BigInt(object.keyMean.toString());
        }
        if (object.keyStdDev !== undefined && object.keyStdDev !== null) {
            message.keyStdDev = BigInt(object.keyStdDev.toString());
        }
        if (object.valueMean !== undefined && object.valueMean !== null) {
            message.valueMean = BigInt(object.valueMean.toString());
        }
        if (object.valueStdDev !== undefined && object.valueStdDev !== null) {
            message.valueStdDev = BigInt(object.valueStdDev.toString());
        }
        if (object.genesisCount !== undefined && object.genesisCount !== null) {
            message.genesisCount = BigInt(object.genesisCount.toString());
        }
        message.insertWeight = object.insertWeight ?? 0;
        message.updateWeight = object.updateWeight ?? 0;
        message.getWeight = object.getWeight ?? 0;
        message.deleteWeight = object.deleteWeight ?? 0;
        return message;
    },
};
