// cosmjs-types/ibc/applications/transfer/v1/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hop = exports.Denom = exports.Token = exports.protobufPackage = void 0;
/* eslint-disable */
const binary_1 = require("../../../../binary");
const helpers_1 = require("../../../../helpers");
exports.protobufPackage = "ibc.applications.transfer.v1";
function createBaseToken() {
    return {
        denom: exports.Denom.fromPartial({}),
        amount: "",
    };
}
exports.Token = {
    typeUrl: "/ibc.applications.transfer.v1.Token",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.denom !== undefined) {
            exports.Denom.encode(message.denom, writer.uint32(10).fork()).ldelim();
        }
        if (message.amount !== "") {
            writer.uint32(18).string(message.amount);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseToken();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.denom = exports.Denom.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.amount = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseToken();
        if ((0, helpers_1.isSet)(object.denom))
            obj.denom = exports.Denom.fromJSON(object.denom);
        if ((0, helpers_1.isSet)(object.amount))
            obj.amount = String(object.amount);
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.denom !== undefined && (obj.denom = message.denom ? exports.Denom.toJSON(message.denom) : undefined);
        message.amount !== undefined && (obj.amount = message.amount);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseToken();
        if (object.denom !== undefined && object.denom !== null) {
            message.denom = exports.Denom.fromPartial(object.denom);
        }
        message.amount = object.amount ?? "";
        return message;
    },
};
function createBaseDenom() {
    return {
        base: "",
        trace: [],
    };
}
exports.Denom = {
    typeUrl: "/ibc.applications.transfer.v1.Denom",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.base !== "") {
            writer.uint32(10).string(message.base);
        }
        for (const v of message.trace) {
            exports.Hop.encode(v, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDenom();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.base = reader.string();
                    break;
                case 3:
                    message.trace.push(exports.Hop.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseDenom();
        if ((0, helpers_1.isSet)(object.base))
            obj.base = String(object.base);
        if (Array.isArray(object?.trace))
            obj.trace = object.trace.map((e) => exports.Hop.fromJSON(e));
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.base !== undefined && (obj.base = message.base);
        if (message.trace) {
            obj.trace = message.trace.map((e) => (e ? exports.Hop.toJSON(e) : undefined));
        }
        else {
            obj.trace = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseDenom();
        message.base = object.base ?? "";
        message.trace = object.trace?.map((e) => exports.Hop.fromPartial(e)) || [];
        return message;
    },
};
function createBaseHop() {
    return {
        portId: "",
        channelId: "",
    };
}
exports.Hop = {
    typeUrl: "/ibc.applications.transfer.v1.Hop",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.portId !== "") {
            writer.uint32(10).string(message.portId);
        }
        if (message.channelId !== "") {
            writer.uint32(18).string(message.channelId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseHop();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.portId = reader.string();
                    break;
                case 2:
                    message.channelId = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseHop();
        if ((0, helpers_1.isSet)(object.portId))
            obj.portId = String(object.portId);
        if ((0, helpers_1.isSet)(object.channelId))
            obj.channelId = String(object.channelId);
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.portId !== undefined && (obj.portId = message.portId);
        message.channelId !== undefined && (obj.channelId = message.channelId);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseHop();
        message.portId = object.portId ?? "";
        message.channelId = object.channelId ?? "";
        return message;
    },
};
