// cosmjs-types/cosmos/epochs/v1beta1/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventEpochStart = exports.EventEpochEnd = exports.protobufPackage = void 0;
/* eslint-disable */
const binary_1 = require("../../../binary");
const helpers_1 = require("../../../helpers");
exports.protobufPackage = "cosmos.epochs.v1beta1";
function createBaseEventEpochEnd() {
    return {
        epochNumber: BigInt(0),
    };
}
exports.EventEpochEnd = {
    typeUrl: "/cosmos.epochs.v1beta1.EventEpochEnd",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.epochNumber !== BigInt(0)) {
            writer.uint32(8).int64(message.epochNumber);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEventEpochEnd();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.epochNumber = reader.int64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseEventEpochEnd();
        if ((0, helpers_1.isSet)(object.epochNumber))
            obj.epochNumber = BigInt(object.epochNumber.toString());
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.epochNumber !== undefined && (obj.epochNumber = (message.epochNumber || BigInt(0)).toString());
        return obj;
    },
    fromPartial(object) {
        const message = createBaseEventEpochEnd();
        if (object.epochNumber !== undefined && object.epochNumber !== null) {
            message.epochNumber = BigInt(object.epochNumber.toString());
        }
        return message;
    },
};
function createBaseEventEpochStart() {
    return {
        epochNumber: BigInt(0),
        epochStartTime: BigInt(0),
    };
}
exports.EventEpochStart = {
    typeUrl: "/cosmos.epochs.v1beta1.EventEpochStart",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.epochNumber !== BigInt(0)) {
            writer.uint32(8).int64(message.epochNumber);
        }
        if (message.epochStartTime !== BigInt(0)) {
            writer.uint32(16).int64(message.epochStartTime);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEventEpochStart();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.epochNumber = reader.int64();
                    break;
                case 2:
                    message.epochStartTime = reader.int64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseEventEpochStart();
        if ((0, helpers_1.isSet)(object.epochNumber))
            obj.epochNumber = BigInt(object.epochNumber.toString());
        if ((0, helpers_1.isSet)(object.epochStartTime))
            obj.epochStartTime = BigInt(object.epochStartTime.toString());
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.epochNumber !== undefined && (obj.epochNumber = (message.epochNumber || BigInt(0)).toString());
        message.epochStartTime !== undefined &&
            (obj.epochStartTime = (message.epochStartTime || BigInt(0)).toString());
        return obj;
    },
    fromPartial(object) {
        const message = createBaseEventEpochStart();
        if (object.epochNumber !== undefined && object.epochNumber !== null) {
            message.epochNumber = BigInt(object.epochNumber.toString());
        }
        if (object.epochStartTime !== undefined && object.epochStartTime !== null) {
            message.epochStartTime = BigInt(object.epochStartTime.toString());
        }
        return message;
    },
};
