// cosmjs-types/ibc/core/client/v2/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.protobufPackage = void 0;
/* eslint-disable */
const binary_1 = require("../../../../binary");
exports.protobufPackage = "ibc.core.client.v2";
function createBaseConfig() {
    return {
        allowedRelayers: [],
    };
}
exports.Config = {
    typeUrl: "/ibc.core.client.v2.Config",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        for (const v of message.allowedRelayers) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.allowedRelayers.push(reader.string());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseConfig();
        if (Array.isArray(object?.allowedRelayers))
            obj.allowedRelayers = object.allowedRelayers.map((e) => String(e));
        return obj;
    },
    toJSON(message) {
        const obj = {};
        if (message.allowedRelayers) {
            obj.allowedRelayers = message.allowedRelayers.map((e) => e);
        }
        else {
            obj.allowedRelayers = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseConfig();
        message.allowedRelayers = object.allowedRelayers?.map((e) => e) || [];
        return message;
    },
};
