// cosmjs-types/ibc/core/client/v2/genesis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisState = exports.GenesisCounterpartyInfo = exports.protobufPackage = void 0;
/* eslint-disable */
const counterparty_1 = require("./counterparty");
const binary_1 = require("../../../../binary");
const helpers_1 = require("../../../../helpers");
exports.protobufPackage = "ibc.core.client.v2";
function createBaseGenesisCounterpartyInfo() {
    return {
        clientId: "",
        counterpartyInfo: counterparty_1.CounterpartyInfo.fromPartial({}),
    };
}
exports.GenesisCounterpartyInfo = {
    typeUrl: "/ibc.core.client.v2.GenesisCounterpartyInfo",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.clientId !== "") {
            writer.uint32(10).string(message.clientId);
        }
        if (message.counterpartyInfo !== undefined) {
            counterparty_1.CounterpartyInfo.encode(message.counterpartyInfo, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisCounterpartyInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.clientId = reader.string();
                    break;
                case 2:
                    message.counterpartyInfo = counterparty_1.CounterpartyInfo.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseGenesisCounterpartyInfo();
        if ((0, helpers_1.isSet)(object.clientId))
            obj.clientId = String(object.clientId);
        if ((0, helpers_1.isSet)(object.counterpartyInfo))
            obj.counterpartyInfo = counterparty_1.CounterpartyInfo.fromJSON(object.counterpartyInfo);
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.clientId !== undefined && (obj.clientId = message.clientId);
        message.counterpartyInfo !== undefined &&
            (obj.counterpartyInfo = message.counterpartyInfo
                ? counterparty_1.CounterpartyInfo.toJSON(message.counterpartyInfo)
                : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGenesisCounterpartyInfo();
        message.clientId = object.clientId ?? "";
        if (object.counterpartyInfo !== undefined && object.counterpartyInfo !== null) {
            message.counterpartyInfo = counterparty_1.CounterpartyInfo.fromPartial(object.counterpartyInfo);
        }
        return message;
    },
};
function createBaseGenesisState() {
    return {
        counterpartyInfos: [],
    };
}
exports.GenesisState = {
    typeUrl: "/ibc.core.client.v2.GenesisState",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        for (const v of message.counterpartyInfos) {
            exports.GenesisCounterpartyInfo.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.counterpartyInfos.push(exports.GenesisCounterpartyInfo.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseGenesisState();
        if (Array.isArray(object?.counterpartyInfos))
            obj.counterpartyInfos = object.counterpartyInfos.map((e) => exports.GenesisCounterpartyInfo.fromJSON(e));
        return obj;
    },
    toJSON(message) {
        const obj = {};
        if (message.counterpartyInfos) {
            obj.counterpartyInfos = message.counterpartyInfos.map((e) => e ? exports.GenesisCounterpartyInfo.toJSON(e) : undefined);
        }
        else {
            obj.counterpartyInfos = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGenesisState();
        message.counterpartyInfos =
            object.counterpartyInfos?.map((e) => exports.GenesisCounterpartyInfo.fromPartial(e)) || [];
        return message;
    },
};
