// cosmjs-types/ibc/core/client/v2/tx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgClientImpl = exports.MsgUpdateClientConfigResponse = exports.MsgUpdateClientConfig = exports.MsgRegisterCounterpartyResponse = exports.MsgRegisterCounterparty = exports.protobufPackage = void 0;
/* eslint-disable */
const config_1 = require("./config");
const binary_1 = require("../../../../binary");
const helpers_1 = require("../../../../helpers");
exports.protobufPackage = "ibc.core.client.v2";
function createBaseMsgRegisterCounterparty() {
    return {
        clientId: "",
        counterpartyMerklePrefix: [],
        counterpartyClientId: "",
        signer: "",
    };
}
exports.MsgRegisterCounterparty = {
    typeUrl: "/ibc.core.client.v2.MsgRegisterCounterparty",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.clientId !== "") {
            writer.uint32(10).string(message.clientId);
        }
        for (const v of message.counterpartyMerklePrefix) {
            writer.uint32(18).bytes(v);
        }
        if (message.counterpartyClientId !== "") {
            writer.uint32(26).string(message.counterpartyClientId);
        }
        if (message.signer !== "") {
            writer.uint32(34).string(message.signer);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgRegisterCounterparty();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.clientId = reader.string();
                    break;
                case 2:
                    message.counterpartyMerklePrefix.push(reader.bytes());
                    break;
                case 3:
                    message.counterpartyClientId = reader.string();
                    break;
                case 4:
                    message.signer = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseMsgRegisterCounterparty();
        if ((0, helpers_1.isSet)(object.clientId))
            obj.clientId = String(object.clientId);
        if (Array.isArray(object?.counterpartyMerklePrefix))
            obj.counterpartyMerklePrefix = object.counterpartyMerklePrefix.map((e) => (0, helpers_1.bytesFromBase64)(e));
        if ((0, helpers_1.isSet)(object.counterpartyClientId))
            obj.counterpartyClientId = String(object.counterpartyClientId);
        if ((0, helpers_1.isSet)(object.signer))
            obj.signer = String(object.signer);
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.clientId !== undefined && (obj.clientId = message.clientId);
        if (message.counterpartyMerklePrefix) {
            obj.counterpartyMerklePrefix = message.counterpartyMerklePrefix.map((e) => (0, helpers_1.base64FromBytes)(e !== undefined ? e : new Uint8Array()));
        }
        else {
            obj.counterpartyMerklePrefix = [];
        }
        message.counterpartyClientId !== undefined && (obj.counterpartyClientId = message.counterpartyClientId);
        message.signer !== undefined && (obj.signer = message.signer);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseMsgRegisterCounterparty();
        message.clientId = object.clientId ?? "";
        message.counterpartyMerklePrefix = object.counterpartyMerklePrefix?.map((e) => e) || [];
        message.counterpartyClientId = object.counterpartyClientId ?? "";
        message.signer = object.signer ?? "";
        return message;
    },
};
function createBaseMsgRegisterCounterpartyResponse() {
    return {};
}
exports.MsgRegisterCounterpartyResponse = {
    typeUrl: "/ibc.core.client.v2.MsgRegisterCounterpartyResponse",
    encode(_, writer = binary_1.BinaryWriter.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgRegisterCounterpartyResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        const obj = createBaseMsgRegisterCounterpartyResponse();
        return obj;
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    fromPartial(_) {
        const message = createBaseMsgRegisterCounterpartyResponse();
        return message;
    },
};
function createBaseMsgUpdateClientConfig() {
    return {
        clientId: "",
        config: config_1.Config.fromPartial({}),
        signer: "",
    };
}
exports.MsgUpdateClientConfig = {
    typeUrl: "/ibc.core.client.v2.MsgUpdateClientConfig",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.clientId !== "") {
            writer.uint32(10).string(message.clientId);
        }
        if (message.config !== undefined) {
            config_1.Config.encode(message.config, writer.uint32(18).fork()).ldelim();
        }
        if (message.signer !== "") {
            writer.uint32(26).string(message.signer);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgUpdateClientConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.clientId = reader.string();
                    break;
                case 2:
                    message.config = config_1.Config.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.signer = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseMsgUpdateClientConfig();
        if ((0, helpers_1.isSet)(object.clientId))
            obj.clientId = String(object.clientId);
        if ((0, helpers_1.isSet)(object.config))
            obj.config = config_1.Config.fromJSON(object.config);
        if ((0, helpers_1.isSet)(object.signer))
            obj.signer = String(object.signer);
        return obj;
    },
    toJSON(message) {
        const obj = {};
        message.clientId !== undefined && (obj.clientId = message.clientId);
        message.config !== undefined && (obj.config = message.config ? config_1.Config.toJSON(message.config) : undefined);
        message.signer !== undefined && (obj.signer = message.signer);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseMsgUpdateClientConfig();
        message.clientId = object.clientId ?? "";
        if (object.config !== undefined && object.config !== null) {
            message.config = config_1.Config.fromPartial(object.config);
        }
        message.signer = object.signer ?? "";
        return message;
    },
};
function createBaseMsgUpdateClientConfigResponse() {
    return {};
}
exports.MsgUpdateClientConfigResponse = {
    typeUrl: "/ibc.core.client.v2.MsgUpdateClientConfigResponse",
    encode(_, writer = binary_1.BinaryWriter.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgUpdateClientConfigResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        const obj = createBaseMsgUpdateClientConfigResponse();
        return obj;
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    fromPartial(_) {
        const message = createBaseMsgUpdateClientConfigResponse();
        return message;
    },
};
class MsgClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
        this.RegisterCounterparty = this.RegisterCounterparty.bind(this);
        this.UpdateClientConfig = this.UpdateClientConfig.bind(this);
    }
    RegisterCounterparty(request) {
        const data = exports.MsgRegisterCounterparty.encode(request).finish();
        const promise = this.rpc.request("ibc.core.client.v2.Msg", "RegisterCounterparty", data);
        return promise.then((data) => exports.MsgRegisterCounterpartyResponse.decode(new binary_1.BinaryReader(data)));
    }
    UpdateClientConfig(request) {
        const data = exports.MsgUpdateClientConfig.encode(request).finish();
        const promise = this.rpc.request("ibc.core.client.v2.Msg", "UpdateClientConfig", data);
        return promise.then((data) => exports.MsgUpdateClientConfigResponse.decode(new binary_1.BinaryReader(data)));
    }
}
exports.MsgClientImpl = MsgClientImpl;
